pkg/portForward/portForward.go

```diff
@@ -0,0 +1,148 @@
+package portForward
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	"github.com/devfile/library/pkg/devfile/parser"
+	parsercommon "github.com/devfile/library/pkg/devfile/parser/data/v2/common"
+
+	"github.com/redhat-developer/odo/pkg/kclient"
+	"github.com/redhat-developer/odo/pkg/libdevfile"
+	"github.com/redhat-developer/odo/pkg/log"
+	"github.com/redhat-developer/odo/pkg/state"
+	"github.com/redhat-developer/odo/pkg/util"
+)
+
+var _ Client = (*PFClient)(nil)
+
+type PFClient struct {
+	kubernetesClient kclient.ClientInterface
+	stateClient      state.Client
+
+	appliedEndpoints map[string][]int
+
+	// stopChan on which to write to stop the port forwarding
+	stopChan chan struct{}
+	// finishedChan is written when the port forwarding is finished
+	finishedChan chan struct{}
+}
+
+func NewPFClient(kubernetesClient kclient.ClientInterface, stateClient state.Client) *PFClient {
+	return &PFClient{
+		kubernetesClient: kubernetesClient,
+		stateClient:      stateClient,
+	}
+}
+
+func (o *PFClient) StartPortForwarding(
+	devFileObj parser.DevfileObj,
+	randomPorts bool,
+	errOut io.Writer,
+) error {
+
+	// get the endpoint/port information for containers in devfile and setup port-forwarding
+	containers, err := devFileObj.Data.GetComponents(parsercommon.DevfileOptions{
+		ComponentOptions: parsercommon.ComponentOptions{ComponentType: v1alpha2.ContainerComponentType},
+	})
+	if err != nil {
+		return err
+	}
+	ceMapping := libdevfile.GetContainerEndpointMapping(containers)
+	if o.stopChan != nil && reflect.DeepEqual(ceMapping, o.appliedEndpoints) {
+		return nil
+	}
+
+	o.appliedEndpoints = ceMapping
```

Comment:
I don't see this variable getting used anywhere. What purpose does it serve?

Reply:
See my previous comment, to check is something has changed in the Devfile's Endpoints since we last started the Port Forwarding